### PR TITLE
[Android] Implement the setInitialScale() API.

### PR DESCRIPTION
**Pull request by wuhengzhi**
This patch is to implement the embedding api for setInitialScale().
Add test case for setInitialScale().

BUG=XWALK-4304, XWALK-4365

### CODE REVIEW

**Comment by wuhengzhi** [comment]
@lincsoon @axinging @fujunwei PTAL, thanks.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkSettings.java [hunk 1]
```diff
@@ -718,6 +721,44 @@ private boolean getSaveFormDataLocked() {
...
+    void setDIPScale(double dipScale) {
+        synchronized (mXWalkSettingsLock) {
+            mDIPScale = dipScale;
+            // TODO(joth): This should also be synced over to native side, but right now
+            // the setDIPScale call is always followed by a setWebContents() which covers this.
+        }
+    }
+
+    /**
+     * See {@link android.webkit.WebView#setInitialScale}.
```
Comment (axinging) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkSettings.java]: How about copy the comments from WebView?

Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkSettings.java]: The comment was documented at https://github.com/wuhengzhi/crosswalk/blob/b7901a51c540a13db49845081dd4797a3f72184b/runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java#L914, do you mean this?

Comment (axinging) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkSettings.java]: It's OK for me.

**Comment by axinging** [comment]
LGTM. @lincsoon  @fujunwei

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java [hunk 2]
```diff
@@ -80,6 +80,9 @@
     long mNativeContent;
     long mNativeWebContents;
 
+    // TODO(boliu): This should be in a global context, not per webview.
+    private double mDIPScale = 1.0;
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: setNativeContent will be always invokde. Why do we need to set the default value here?

**Comment by wuhengzhi** [comment]
@lincsoon Done, thanks.

**Comment by wuhengzhi** [comment]
@fujunwei Done, thanks.

**Comment by fujunwei** [comment]
LGTM.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java [hunk 3]
```diff
@@ -80,6 +80,9 @@
     long mNativeContent;
     long mNativeWebContents;
 
+    // TODO(hengzhi.wu): This should be in a global context, not per XWalkView.
+    private double mDIPScale;
```
Comment (halton) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: please follow chromium naming rule, `DIP` -> `Dip`, please apply in all places in this PR.

Comment (axinging) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: I suggest not. DIP means device independent pixels. Not a English word dip. Also, android webview has such a style.

File: runtime/browser/android/xwalk_settings.cc [hunk 4]
```diff
@@ -269,6 +270,23 @@ static jstring GetDefaultUserAgent(JNIEnv* env, jclass clazz) {
...
+  float initial_page_scale_percent =
+      Java_XWalkSettings_getInitialPageScalePercentLocked(env, obj);
+  if (initial_page_scale_percent == 0) {
+    render_view_host_ext->SetInitialPageScale(-1);
+  } else {
+    float dip_scale = static_cast<float>(
+        Java_XWalkSettings_getDIPScaleLocked(env, obj));
+    render_view_host_ext->SetInitialPageScale(
+        initial_page_scale_percent / dip_scale / 100.0f);
+  }
```
Comment (halton) [inline runtime/browser/android/xwalk_settings.cc]: nit, rewrite as below to save indent in else block
```
if (...) {
  render_view_host_ext->SetInitialPageScale(-1);
  return;
}

...
```

File: test/android/core/javatests/src/org/xwalk/core/xwview/test/XWalkViewTestBase.java [hunk 5]
```diff
@@ -663,4 +664,31 @@ public void run() {
...
+        });
+    }
+
+    protected double getDipScale() {
+        return DeviceDisplayInfo.create(mXWalkView.getContext()).getDIPScale();
+    }
+
+    protected float getScaleFactor() {
+        float scale = mTestHelperBridge.getOnScaleChangedHelper().getNewScale();
+        return scale / (float)getDipScale();
```
Comment (halton) [inline test/android/core/javatests/src/org/xwalk/core/xwview/test/XWalkViewTestBase.java]: space after (float)

File: test/android/core/javatests/src/org/xwalk/core/xwview/test/SetInitialScaleTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        final double dipScale = getDipScale();
+
+        // Page scale updates are asynchronous. There is an issue that we can't
+        // reliably check, whether the scale as NOT changed (i.e. remains to be 1.0).
+        // So we first change the scale to some non-default value, and then wait
+        // until it gets back to 1.0.
+        int onScaleChangedCallCount = mOnScaleChangedHelper.getCallCount();
+        loadDataSync(null, pageScale4, "text/html", false);
+        mOnScaleChangedHelper.waitForCallback(onScaleChangedCallCount);
+        float scale = (float)dipScale * initialScale4;
```
Comment (halton) [inline test/android/core/javatests/src/org/xwalk/core/xwview/test/SetInitialScaleTest.java]: space after (float)

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java (hunk 3 above)
Comment (halton) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: Okay

**Comment by axinging** [comment]
@wuhengzhi, please help to make this happen.

**Comment by darktears** [comment]
@kenchris

**Comment by wuhengzhi** [comment]
The trybot seems not to be triggered.

**Comment by halton** [comment]
@wuhengzhi if you want to trybot to run again, amend locally w/o any change and force update.

**Comment by wuhengzhi** [comment]
I just rebased this PR to resolve conflict, but the trybot was also not triggered.

**Comment by halton** [comment]
ping @rakuco for trybot issue.

**Comment by wuhengzhi** [comment]
@halton The bots are green, PTAL again, thanks.

**Comment by halton** [comment]
LGTM, merging.

**Pull request merged**